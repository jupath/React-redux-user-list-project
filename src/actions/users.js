
export const usersAreLoading = bool => ({
  type: 'USERS_ARE_LOADING',
  isLoading: bool,
});

export const setUsers = users => ({
  type: 'SET_USERS',
  users,
});

export const startSetUsers = url => async (dispatch) => {
  dispatch(usersAreLoading(true));

  try {
    const response = await fetch(url);
    const data = await response.json();
    const results = await data.results;
    const users = await results.map(user => ({
      username: user.login.username,
      firstName: user.name.first,
      lastName: user.name.last,
      email: user.email,
      address: `${user.location.street}, ${user.location.city}, ${user.location.state}, ${user.location.postcode}`,
      dob: user.dob,
      cell: user.cell,
      image: user.picture.large,
    }));
    await dispatch(usersAreLoading(false));
    await dispatch(setUsers(users));
  } catch (e) {
    console.error(e);
  }
};

export const updateUser = (username, updatedData) => ({
  type: 'UPDATE_USER',
  username,
  updatedData,
});
