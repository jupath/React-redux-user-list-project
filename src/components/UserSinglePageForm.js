import React, { Component } from 'react';
import AutosizeInput from 'react-input-autosize';
import { Button } from 'reactstrap';
import PropTypes from 'prop-types';

class UserSinglePageForm extends Component {
  state = {
    text: this.props.data,
  }

  updateState = (event) => {
    const text = event.target.value;
    this.setState({
      text,
    });
  }

  submitForm = (event) => {
    event.preventDefault();
    const { text } = this.state;
    this.props.handleUpdateUser(text);
  }

  render() {
    return (
      <form onSubmit={this.submitForm} className="user-single__form">
        <AutosizeInput
          type="text"
          value={this.state.text}
          onChange={this.updateState}
        />
        <Button color="success">save</Button>
      </form>
    );
  }
}

UserSinglePageForm.propTypes = {
  data: PropTypes.string.isRequired,
  handleUpdateUser: PropTypes.func.isRequired,
};

export default UserSinglePageForm;
