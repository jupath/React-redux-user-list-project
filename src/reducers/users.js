export const usersAreLoading = (state = false, action) => {
  switch (action.type) {
    case 'USERS_ARE_LOADING':
      return action.isLoading;
    default:
      return state;
  }
};

export const users = (state = [], action) => {
  switch (action.type) {
    case 'SET_USERS':
      return action.users;
    case 'UPDATE_USER':
      return state.map((user) => {
        if (user.username === action.username) {
          return {
            ...user,
            ...action.updatedData,
          };
        }
        return user;
      });
    default:
      return state;
  }
};
