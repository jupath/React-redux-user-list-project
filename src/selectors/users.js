
export default (users, { text, orderBy }) => {
  return users.filter(user => {
    const regex = new RegExp( text, 'gi' );
    return user.firstName.match( regex ) || user.lastName.match( regex );
  }).sort((a, b) => {
    if( orderBy === 'desc' ) {
      return a.lastName < b.lastName ? 1 : -1;
    } else if ( orderBy === 'asc' ) {
      return a.lastName < b.lastName ? -1 : 1;
    }
  });
}