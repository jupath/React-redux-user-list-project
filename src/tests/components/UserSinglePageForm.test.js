import React from 'react';
import { shallow } from 'enzyme';
import UserSinglePageForm from '../../components/UserSinglePageForm';
import users from '../fixtures/users';

test('should render UserSinglePageForm correctly', () => {
  const wrapper = shallow(<UserSinglePageForm data={users[0].email} />);
  expect(wrapper).toMatchSnapshot();
});

test('should set text in input change', () => {
  const value = 'test@example.com';
  const wrapper = shallow(<UserSinglePageForm data={users[0].email} />);
  wrapper.find('AutosizeInput').simulate('change', {
    target: { value },
  });
  expect(wrapper.state('text')).toBe(value);
});

test('should handle submit form correctly', () => {
  const value = 'test@example.com';
  const handleUpdateUserSpy = jest.fn();
  const wrapper = shallow(<UserSinglePageForm data={value} handleUpdateUser={handleUpdateUserSpy} />);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { },
  });
  expect(handleUpdateUserSpy).toHaveBeenLastCalledWith(value);
});
