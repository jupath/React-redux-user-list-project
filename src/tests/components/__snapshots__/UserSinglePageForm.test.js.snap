// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render UserSinglePageForm correctly 1`] = `
<form
  className="user-single__form"
  onSubmit={[Function]}
>
  <AutosizeInput
    injectStyles={true}
    minWidth={1}
    onChange={[Function]}
    type="text"
    value="romain.hoogmoed@example.com"
  />
  <Button
    color="success"
    tag="button"
  >
    save
  </Button>
</form>
`;
